--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -2,6 +_,8 @@
 
 import java.util.function.Function;
 import java.util.function.ToIntFunction;
+
+import net.fabricmc.fabric.api.event.registry.RegistryEntryAddedCallback;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Registry;
@@ -7235,5 +_,11 @@
                 blockState.initCache();
             }
         }
+        // Ensure that any blocks added after this point have their shape cache initialized.
+        RegistryEntryAddedCallback.event(BuiltInRegistries.BLOCK).register((rawId, id, block) -> {
+            for (BlockState state : block.getStateDefinition().getPossibleStates()) {
+                state.initCache();
+            }
+        });
     }
 }
