--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -42,6 +_,8 @@
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.level.material.MapColor;
 import net.minecraft.world.level.material.PushReaction;
+import org.bukkit.Bukkit;
+import org.teneted.tenet.event.RegisterEvents;
 
 public class Blocks {
     private static final BlockBehaviour.StatePredicate NOT_CLOSED_SHULKER = (state, level, pos) -> !(
@@ -7209,6 +_,7 @@
 
     private static Block register(ResourceKey<Block> resourceKey, Function<BlockBehaviour.Properties, Block> factory, BlockBehaviour.Properties properties) {
         Block block = factory.apply(properties.setId(resourceKey));
+        RegisterEvents.REGISTER_BLOCK.invoker().register();
         return Registry.register(BuiltInRegistries.BLOCK, resourceKey, block);
     }
 
