--- a/net/minecraft/core/Registry.java
+++ b/net/minecraft/core/Registry.java
@@ -14,6 +_,8 @@
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 import javax.annotation.Nullable;
+
+import net.fabricmc.fabric.api.event.registry.FabricRegistry;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.tags.TagKey;
@@ -21,7 +_,7 @@
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
 
-public interface Registry<T> extends Keyable, HolderLookup.RegistryLookup<T>, IdMap<T> {
+public interface Registry<T> extends Keyable, HolderLookup.RegistryLookup<T>, IdMap<T>, FabricRegistry {
     @Override
     ResourceKey<? extends Registry<T>> key();
 
