--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -28,6 +_,9 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
+
+import net.fabricmc.fabric.api.event.player.PlayerPickItemEvents;
+import net.fabricmc.fabric.api.event.player.UseEntityCallback;
 import net.minecraft.ChatFormatting;
 import net.minecraft.Util;
 import net.minecraft.advancements.AdvancementHolder;
@@ -205,6 +_,7 @@
 import net.minecraft.world.level.storage.TagValueOutput;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.BlockHitResult;
+import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.BooleanOp;
 import net.minecraft.world.phys.shapes.Shapes;
@@ -1013,7 +_,7 @@
             if (serverLevel.isLoaded(blockPos)) {
                 BlockState blockState = serverLevel.getBlockState(blockPos);
                 boolean flag = this.player.hasInfiniteMaterials() && packet.includeData();
-                ItemStack cloneItemStack = blockState.getCloneItemStack(serverLevel, blockPos, flag);
+                ItemStack cloneItemStack = onPickItemFromBlock(blockState, blockPos, blockState.getCloneItemStack(serverLevel, blockPos, flag), packet);
                 if (!cloneItemStack.isEmpty()) {
                     if (flag && this.player.getBukkitEntity().hasPermission("minecraft.nbt.copy")) { // Spigot
                         addBlockDataToItem(blockState, serverLevel, blockPos, cloneItemStack);
@@ -1025,6 +_,18 @@
         }
     }
 
+    public ItemStack onPickItemFromBlock(BlockState state, BlockPos pos, ItemStack original, ServerboundPickItemFromBlockPacket packet) {
+        ItemStack stack = PlayerPickItemEvents.BLOCK.invoker().onPickItemFromBlock(player, pos, state, packet.includeData());
+        if (stack == null) {
+            return original;
+        } else if (!stack.isEmpty()) {
+            this.tryPickItem(stack, pos, null, packet.includeData());
+        }
+
+        // Prevent vanilla data-inclusion behavior
+        return ItemStack.EMPTY;
+    }
+
     private static void addBlockDataToItem(BlockState state, ServerLevel level, BlockPos pos, ItemStack stack) {
         BlockEntity blockEntity = state.hasBlockEntity() ? level.getBlockEntity(pos) : null;
         if (blockEntity != null) {
@@ -1044,13 +_,26 @@
         PacketUtils.ensureRunningOnSameThread(packet, this, serverLevel);
         Entity entityOrPart = serverLevel.getEntityOrPart(packet.id());
         if (entityOrPart != null && this.player.canInteractWithEntity(entityOrPart, 3.0)) {
-            ItemStack pickResult = entityOrPart.getPickResult();
+            ItemStack pickResult = onPickItemFromEntity(entityOrPart, entityOrPart.getPickResult(), packet);
             if (pickResult != null && !pickResult.isEmpty()) {
                 this.tryPickItem(pickResult, null, entityOrPart, packet.includeData()); // Paper - Extend PlayerPickItemEvent API
             }
         }
     }
 
+    public ItemStack onPickItemFromEntity(Entity entity, ItemStack original, ServerboundPickItemFromEntityPacket packet) {
+        ItemStack stack = PlayerPickItemEvents.ENTITY.invoker().onPickItemFromEntity(player, entity, packet.includeData());
+
+        if (stack == null) {
+            return original;
+        } else if (!stack.isEmpty()) {
+            this.tryPickItem(stack, null, entity, packet.includeData());
+        }
+
+        // Prevent vanilla data-inclusion behavior
+        return ItemStack.EMPTY;
+    }
+
     private void tryPickItem(ItemStack stack, @Nullable BlockPos blockPos, @Nullable Entity entity, boolean includeData) { // Paper - Extend PlayerPickItemEvent API
         if (stack.isItemEnabled(this.player.level().enabledFeatures())) {
             Inventory inventory = this.player.getInventory();
@@ -2826,11 +_,21 @@
 
                             @Override
                             public void onInteraction(InteractionHand hand) {
+                                InteractionResult result =  UseEntityCallback.EVENT.invoker().interact(player, player.level(), hand, target, null);
+                                if (result != InteractionResult.PASS) {
+                                    return;
+                                }
                                 this.performInteraction(hand, Player::interactOn, new PlayerInteractEntityEvent(ServerGamePacketListenerImpl.this.getCraftPlayer(), target.getBukkitEntity(), org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand))); // CraftBukkit
                             }
 
                             @Override
                             public void onInteraction(InteractionHand hand, Vec3 interactionLocation) {
+                                EntityHitResult hitResult = new EntityHitResult(target, interactionLocation.add(target.getX(), target.getY(), target.getZ()));
+                                InteractionResult result = UseEntityCallback.EVENT.invoker().interact(player, player.level(), hand, target, hitResult);
+
+                                if (result != InteractionResult.PASS) {
+                                    return;
+                                }
                                 this.performInteraction(
                                     hand, (player, entity, interactionHand) -> entity.interactAt(player, interactionLocation, interactionHand), new PlayerInteractAtEntityEvent(ServerGamePacketListenerImpl.this.getCraftPlayer(), target.getBukkitEntity(), org.bukkit.craftbukkit.util.CraftVector.toBukkit(interactionLocation), org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand)) // CraftBukkit
                                 );
