--- a/net/minecraft/server/Bootstrap.java
+++ b/net/minecraft/server/Bootstrap.java
@@ -55,11 +_,7 @@
                     EntitySelectorOptions.bootStrap();
                     DispenseItemBehavior.bootStrap();
                     CauldronInteraction.bootStrap();
-                    // Paper start
-                    BuiltInRegistries.bootStrap(() -> {
-                        io.papermc.paper.world.worldgen.OptionallyFlatBedrockConditionSource.bootstrap(); // Paper - Flat bedrock generator settings
-                    });
-                    // Paper end
+                    BuiltInRegistries.createContents();
                     CreativeModeTabs.validate();
                     wrapStreams();
                     bootstrapDuration.set(Duration.between(instant, Instant.now()).toMillis());
