--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -5,6 +_,8 @@
 import java.util.HashSet;
 import java.util.Set;
 import javax.annotation.Nullable;
+
+import net.fabricmc.fabric.api.blockview.v2.RenderDataBlockEntity;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.HolderLookup;
@@ -36,7 +_,7 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import org.slf4j.Logger;
 
-public abstract class BlockEntity implements DebugValueSource {
+public abstract class BlockEntity implements DebugValueSource, RenderDataBlockEntity {
     static boolean ignoreBlockEntityUpdates; // Paper - Perf: Optimize Hoppers
     // CraftBukkit start - data containers
     private static final org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry();
