--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -30,6 +_,9 @@
 import java.util.function.Predicate;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+
+import net.fabricmc.fabric.api.block.v1.BlockFunctionalityTags;
+import net.fabricmc.fabric.api.event.lifecycle.v1.ServerEntityEvents;
 import net.minecraft.BlockUtil;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -2181,6 +_,18 @@
             return false;
         } else {
             BlockState blockState = this.level().getBlockState(pos.below());
+            if (blockState.is(BlockFunctionalityTags.CAN_CLIMB_TRAPDOOR_ABOVE)) {
+                if (blockState.getBlock() instanceof LadderBlock) {
+                    // Check that the ladder and trapdoor are placed in the same direction.
+                    // Vanilla does the same check for the normal ladder block.
+                    if (blockState.getValue(LadderBlock.FACING) == state.getValue(TrapDoorBlock.FACING)) {
+                        return true;
+                    }
+                } else {
+                    // Don't do any checks for other blocks. They might not even have the facing property.
+                    return true;
+                }
+            }
             return blockState.is(Blocks.LADDER) && blockState.getValue(LadderBlock.FACING) == state.getValue(TrapDoorBlock.FACING);
         }
     }
@@ -3399,7 +_,7 @@
                     map = Maps.newEnumMap(EquipmentSlot.class);
                     equipmentChanges = Maps.newEnumMap(org.bukkit.inventory.EquipmentSlot.class); // Paper - EntityEquipmentChangedEvent
                 }
-
+                ServerEntityEvents.EQUIPMENT_CHANGE.invoker().onChange(this, equipmentSlot, itemStack, itemBySlot);
                 map.put(equipmentSlot, itemBySlot);
                 equipmentChanges.put(org.bukkit.craftbukkit.CraftEquipmentSlot.getSlot(equipmentSlot), new EquipmentChangeImpl(oldItem, newItem)); // Paper - EntityEquipmentChangedEvent
                 AttributeMap attributes = this.getAttributes();
