--- a/net/minecraft/core/registries/BuiltInRegistries.java
+++ b/net/minecraft/core/registries/BuiltInRegistries.java
@@ -154,7 +_,7 @@
 public class BuiltInRegistries {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final Map<ResourceLocation, Supplier<?>> LOADERS = Maps.newLinkedHashMap();
-    private static final WritableRegistry<WritableRegistry<?>> WRITABLE_REGISTRY = new MappedRegistry<>(
+    private static WritableRegistry<WritableRegistry<?>> WRITABLE_REGISTRY = new MappedRegistry<>(
         ResourceKey.createRegistryKey(Registries.ROOT_REGISTRY_NAME), Lifecycle.stable()
     );
     public static final DefaultedRegistry<GameEvent> GAME_EVENT = registerDefaulted(Registries.GAME_EVENT, "step", GameEvent::bootstrap);
@@ -341,6 +_,11 @@
         new net.minecraft.resources.RegistryOps.HolderLookupAdapter(net.minecraft.server.RegistryLayer.STATIC_ACCESS)
     );
     // Paper end - add built-in registry conversions
+    private static boolean hasInitialised = false;
+
+    public static WritableRegistry<WritableRegistry<?>> getWRITABLE_REGISTRY() {
+        return WRITABLE_REGISTRY;
+    }
 
     private static <T> Registry<T> registerSimple(ResourceKey<? extends Registry<T>> registryKey, BuiltInRegistries.RegistryBootstrap<T> bootstrap) {
         return internalRegister(registryKey, new MappedRegistry<>(registryKey, Lifecycle.stable(), false), bootstrap);
@@ -388,7 +_,12 @@
         validate(REGISTRY);
     }
 
-    private static void createContents() {
+    public static void createContents() {
+        if (hasInitialised) {
+            return;
+        }
+
+        hasInitialised = true;
         // Paper start - class-load org.bukkit.Registry
         // we have to class-load Registry here to create all the CraftRegistry instances
         // that will be created when Registry is class-loaded before RegistryAccess#getRegistry
