--- a/net/minecraft/core/registries/Registries.java
+++ b/net/minecraft/core/registries/Registries.java
@@ -295,10 +_,21 @@
     }
 
     public static String elementsDirPath(ResourceKey<? extends Registry<?>> registryKey) {
+        ResourceLocation id = registryKey.registry();
+
+        if (!id.getNamespace().equals(ResourceLocation.DEFAULT_NAMESPACE)) {
+            return id.getNamespace() + "/" + id.getPath();
+        }
+
         return registryKey.location().getPath();
     }
 
     public static String tagsDirPath(ResourceKey<? extends Registry<?>> registryKey) {
+        ResourceLocation id = registryKey.registry();
+
+        if (!id.getNamespace().equals(ResourceLocation.DEFAULT_NAMESPACE)) {
+            return "tags/" + id.getNamespace() + "/" + id.getPath();
+        }
         return "tags/" + registryKey.location().getPath();
     }
 }
