--- a/net/minecraft/core/registries/Registries.java
+++ b/net/minecraft/core/registries/Registries.java
@@ -294,11 +_,28 @@
         return ResourceKey.createRegistryKey(ResourceLocation.withDefaultNamespace(name));
     }
 
+    // Vanilla doesn't mark namespaces in the directories of tags and dynamic registry elements at all,
+    // so we prepend the directories with the namespace if it's a modded registry id.
     public static String elementsDirPath(ResourceKey<? extends Registry<?>> registryKey) {
-        return registryKey.location().getPath();
+        String original = registryKey.location().getPath();
+        ResourceLocation id = registryKey.registry();
+
+        if (!id.getNamespace().equals(ResourceLocation.DEFAULT_NAMESPACE)) {
+            return id.getNamespace() + "/" + id.getPath();
+        }
+        return original;
     }
 
+    // Vanilla doesn't mark namespaces in the directories of tags and dynamic registry elements at all,
+    // so we prepend the directories with the namespace if it's a modded registry id.
     public static String tagsDirPath(ResourceKey<? extends Registry<?>> registryKey) {
-        return "tags/" + registryKey.location().getPath();
+        String original = "tags/" + registryKey.location().getPath();
+        ResourceLocation id = registryKey.registry();
+
+        if (!id.getNamespace().equals(ResourceLocation.DEFAULT_NAMESPACE)) {
+            return "tags/" + id.getNamespace() + "/" + id.getPath();
+        }
+
+        return original;
     }
 }
