--- a/net/minecraft/data/registries/RegistryPatchGenerator.java
+++ b/net/minecraft/data/registries/RegistryPatchGenerator.java
@@ -3,6 +_,8 @@
 import com.mojang.datafixers.DataFixUtils;
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
+
+import net.fabricmc.fabric.api.event.registry.DynamicRegistries;
 import net.minecraft.core.Cloner;
 import net.minecraft.core.HolderLookup;
 import net.minecraft.core.RegistryAccess;
@@ -21,6 +_,7 @@
             provider -> {
                 RegistryAccess.Frozen frozen = RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY);
                 Cloner.Factory factory = new Cloner.Factory();
+                RegistryDataLoader.WORLDGEN_REGISTRIES = DynamicRegistries.getDynamicRegistries();
                 RegistryDataLoader.WORLDGEN_REGISTRIES.forEach(registryData -> registryData.runWithArguments(factory::addCodec));
                 RegistrySetBuilder.PatchedRegistries patchedRegistries = registrySetBuilder.buildPatch(frozen, provider, factory);
                 HolderLookup.Provider provider1 = patchedRegistries.full();
