--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -42,6 +_,12 @@
                         .defaultsTo(new File("plugins"))
                         .describedAs("Plugin directory");
 
+                this.acceptsAll(asList("M", "mods"), "Mod directory to use")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("mods"))
+                        .describedAs("Mod directory");
+
                 this.acceptsAll(asList("h", "host", "server-ip"), "Host to listen on")
                         .withRequiredArg()
                         .ofType(String.class)
