--- a/net/minecraft/world/level/levelgen/DebugLevelSource.java
+++ b/net/minecraft/world/level/levelgen/DebugLevelSource.java
@@ -34,11 +_,11 @@
         instance -> instance.group(RegistryOps.retrieveElement(Biomes.PLAINS)).apply(instance, instance.stable(DebugLevelSource::new))
     );
     private static final int BLOCK_MARGIN = 2;
-    private static final List<BlockState> ALL_BLOCKS = StreamSupport.stream(BuiltInRegistries.BLOCK.spliterator(), false)
+    private static List<BlockState> ALL_BLOCKS = StreamSupport.stream(BuiltInRegistries.BLOCK.spliterator(), false)
         .flatMap(block -> block.getStateDefinition().getPossibleStates().stream())
         .collect(Collectors.toList());
-    private static final int GRID_WIDTH = Mth.ceil(Mth.sqrt(ALL_BLOCKS.size()));
-    private static final int GRID_HEIGHT = Mth.ceil((float)ALL_BLOCKS.size() / GRID_WIDTH);
+    private static int GRID_WIDTH = Mth.ceil(Mth.sqrt(ALL_BLOCKS.size()));
+    private static int GRID_HEIGHT = Mth.ceil((float)ALL_BLOCKS.size() / GRID_WIDTH);
     protected static final BlockState AIR = Blocks.AIR.defaultBlockState();
     protected static final BlockState BARRIER = Blocks.BARRIER.defaultBlockState();
     public static final int HEIGHT = 70;
@@ -51,6 +_,18 @@
     @Override
     protected MapCodec<? extends ChunkGenerator> codec() {
         return CODEC;
+    }
+
+    public static void setALL_BLOCKS(List<BlockState> blockStates) {
+        ALL_BLOCKS = blockStates;
+    }
+
+    public static void setGRID_WIDTH(int length) {
+        GRID_WIDTH = length;
+    }
+
+    public static void setGRID_HEIGHT(int length) {
+        GRID_HEIGHT = length;
     }
 
     @Override
