--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -19,6 +_,8 @@
 import java.util.function.Function;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents;
 import net.minecraft.ChatFormatting;
 import net.minecraft.FileUtil;
 import net.minecraft.commands.CommandSourceStack;
@@ -65,6 +_,7 @@
 import net.minecraft.server.PlayerAdvancements;
 import net.minecraft.server.RegistryLayer;
 import net.minecraft.server.ServerScoreboard;
+import net.minecraft.server.dedicated.DedicatedServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.network.CommonListenerCookie;
@@ -98,6 +_,7 @@
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.Team;
 import org.slf4j.Logger;
+import org.teneted.tenet.init.BukkitRegistry;
 
 public abstract class PlayerList {
     public static final File USERBANLIST_FILE = new File("banned-players.json");
@@ -143,6 +_,7 @@
         this.playerIo = playerIo;
         this.whitelist = new UserWhiteList(WHITELIST_FILE, notificationService);
         this.ops = new ServerOpList(OPLIST_FILE, notificationService);
+        BukkitRegistry.registerAll(((DedicatedServer) server));
         this.bans = new UserBanList(USERBANLIST_FILE, notificationService);
         this.ipBans = new IpBanList(IPBANLIST_FILE, notificationService);
     }
@@ -190,6 +_,7 @@
         serverGamePacketListenerImpl.send(new ClientboundPlayerAbilitiesPacket(player.getAbilities()));
         serverGamePacketListenerImpl.send(new ClientboundSetHeldSlotPacket(player.getInventory().getSelectedSlot()));
         RecipeManager recipeManager = this.server.getRecipeManager();
+        ServerLifecycleEvents.SYNC_DATA_PACK_CONTENTS.invoker().onSyncDataPackContents(player, true);
         serverGamePacketListenerImpl.send(
             new ClientboundUpdateRecipesPacket(recipeManager.getSynchronizedItemProperties(), recipeManager.getSynchronizedStonecutterRecipes())
         );
@@ -1262,6 +_,9 @@
         // Paper start - API for updating recipes on clients
     }
     public void reloadTagData() {
+        for (ServerPlayer player : this.getPlayers()) {
+            ServerLifecycleEvents.SYNC_DATA_PACK_CONTENTS.invoker().onSyncDataPackContents(player, false);
+        }
         this.broadcastAll(new ClientboundUpdateTagsPacket(TagNetworkSerialization.serializeTagsToNetwork(this.registries)));
         // CraftBukkit start
         // this.reloadRecipes(); // Paper - do not reload recipes just because tag data was reloaded
